### ISSUE TITLE
[MU4 Issue] Ossia issues (cutaway staff)

### ISSUE THREAD

**Issue by SilverGreen93** [musescore/MuseScore#15600]
**Describe the bug**
There are a number of issues with cutaway staves that should be pointed out.
The following examples assume that there is a Piano grand staff and you add an extra staff and make it narrow and cutaway.

### 1. The barline cannot be different from the main staff.
If I change the barline of an ossia measure to dotted, dash or any kind of barline, the main staff barline changes as well, even though the selected barline/measure is just the upper one.
This is annoying as I would like to keep the main barline a solid line, but make only the ossia barline dotted in my case.
![image](https://user-images.githubusercontent.com/24845602/209541065-e03f7fe3-d4c3-427c-b0ab-0c1c3f33691e.png)

### 2. Adding an ossia staff adds space between all systems in all pages.
If I add just two visible measures on the ossia staff, the whole layout changes to accomodate an invisible staff, although there are only 2 measures visible in only one system.
There should be no space added between the systems that do not contain any visible measures in the ossia staff.
![image](https://user-images.githubusercontent.com/24845602/209541559-685b5b52-17aa-427c-9ef2-22c863da0d9e.png)

### 2a. Adding an ossia staff in the middle is a more visible issue of (2)
(Just a consequence of 2)
The spacing between the two staves of the Grand staff should not change where there is no ossia staff visible.
![image](https://user-images.githubusercontent.com/24845602/209541706-ac0775f9-6b8e-467f-ba19-d007d2f3452d.png)

### 3. Adding an ossia staff last, keeps the system bracket over the invisible staff.
Although adding the staff first makes the system bracket only cover the reamining 2 staves (which is expected behaviour), adding it as the last staff makes the system bracket go over the invisible ossia staff for no reason.
![image](https://user-images.githubusercontent.com/24845602/209541863-cf824530-a9da-4a81-9e5c-176581012321.png)

**Comment by oktophonie** [comment]
1. I would consider a feature request (the ability to have differing barlines on certain staves), which has been requested before and we'll certainly look into.

2. and 2a. can be avoided by ensuring 'Hide empty staves' is checked and that 'Don't hide empty staves in first system' is unchecked. I have no objection to the idea that if all of the bars in a given system are cutaway are not displayed, then that stave should be considered as invisible - except that this would make these bars unselectable, for when you wished to show them.

(Extra options are needed for stave hiding to make this workaround more viable, for example, options to exclude staves from being hidden, or to ensure that if one stave of a grand staff instrument is shown, the other must be too, etc; this is in the works.)

3. is possible and will require some investigation.

**Comment by MarcSabatella** [comment]
Maybe I'm misunderstanding, but different barlines for different staves *is* supposed - by adding while holding Ctrl.  Same method used for local key and time signatures.

Regarding 2, you can also simply set "Hide when empty" to "Always" in staff/part properties, instead of needing to set hide empty staves for the whole score.  Useful if there are other empty staves you don't hidden.  Realistically, we could have made the "cutaway" option *automatically* hide this staff when empty, as there probably are few if any use cases where you *wouldn't* want this.  But I guarantee 10 minutes after someone implements that someone will complain, "hey, I was using that!" :-)

**Comment by SilverGreen93** [comment]
@MarcSabatella You're right, holding Control while adding the measure barline works. It does the trick and the barline can now be split into different types.
![image](https://user-images.githubusercontent.com/24845602/209577308-41291cb7-2af1-4e3d-8ff6-1c61a5506b83.png)

@MarcSabatella, @oktophonie 
I tried the Hide when empty set to always, but I think that unfortunatelly I found a bug here as well:
On the same example as above, if I go into the staff's properties and set **Hide when empty to always**, immediately after clicking OK, the clefs for the other staves are replaced with either treble & treble (instead of treble & bass) or bass & treble (depending where the ossia staff was in the stack of staves for the instrument). 
I think something messes up the clef's index or associations with the staves
![image](https://user-images.githubusercontent.com/24845602/209577620-0386dae2-3f95-4dad-a23f-ac093ec402c3.png)

Here is the file with the wrong clefs: [ossia2.zip](https://github.com/musescore/MuseScore/files/10304573/ossia2.zip)

Other than this bug, this is however the way I want it to work. Selecting hide when empty results in no more spacing added between systems.

**Comment by Enrico9494** [comment]
Hi team, @oktophonie @MarcSabatella  - has there been any progress on point 3. raised by the original poster @SilverGreen93 please?

I've tried various options for the system bracket, but none seem to produce the desired effect.  I can make it one stave high, but MS chooses the ossia stave and not the main one - and won't allow me to change it.

Many thanks,
Eric :-)

**Comment by Enrico9494** [comment]
I should have said, I'm using MS 4.1.1 Appimage for Linux/Ubuntu 22.04

**Comment by sammik** [comment]
@Enrico9494 this?

https://github.com/musescore/MuseScore/assets/1646034/afc909e3-cdd0-40b9-bd1e-28f7a1dd48c9

**Comment by mccoydtromb** [comment]
> 2. and 2a. can be avoided by ensuring 'Hide empty staves' is checked and that 'Don't hide empty staves in first system' is unchecked. I have no objection to the idea that if all of the bars in a given system are cutaway are not displayed, then that stave should be considered as invisible - except that this would make these bars unselectable, for when you wished to show them. (Extra options are needed for stave hiding to make this workaround more viable, for example, options to exclude staves from being hidden, or to ensure that if one stave of a grand staff instrument is shown, the other must be too, etc; this is in the works.)

As a long time user of MuseScore, I made a similar mistake, and assumed that "Hide all bars ..." would hide the bars and the space they take up. I would suggest either making the UI automatically select the Always Hide button when "Hide all bars..." is selected (which would still allow those few users to change it back, Marc!), or else remove the word "Hide" in favour of cutaway only. Glad I know now, but I found various workarounds before finding this thread showing how it should really be used.
